File: ds_9_bit_operation/L371-m.cpp
//
// Created by 李恒 on 2020/6/29.
//
#include <iostream>

using namespace std;
class Solution {
public:
    int getSum(int a, int b) {
        while(b){
            auto carry = ((unsigned int)(a&b))<<1;
            a = a^b;
            b = carry;
        }
        return a;
    }
};

// a ^ b可以得到两数相加不进位的加法结果
// (a & b) << 1可以得到两数相加产生的进位
//当a & b的结果是负数时，左移就会造成符号位的溢出，所以此处需要转换为unsigned int来避免可能出现的左移越界行为。

File: python-algorithm/ds_9_bit_operation/L371-m.py
# -*- coding: utf-8 -*-


# a ^ b可以得到两数相加不进位的加法结果
# (a & b) << 1可以得到两数相加产生的进位
class Solution:
    def getSum(self, a: int, b: int) -> int:
        while b:
            tmp = (a&b) << 1
            a = a^b
            b = tmp
        return a